File: YunLog/src/main/java/com/example/yunlog/java/YunLogStackTraceFormatter.java
package com.example.yunlog.java;

class YunLogStackTraceFormatter implements YunLogFormatter<StackTraceElement[]> {
    @Override
    public String format(StackTraceElement[] stackTraceElements) {
        StringBuffer buffer = new StringBuffer(128);

        if (stackTraceElements.length == 0) {
            return null;
        } else if (stackTraceElements.length == 1) {
            return stackTraceElements[0].toString();
        } else {
            for (int i = 0, len = stackTraceElements.length; i < len; i++) {
                if (i == 0) {
                    buffer.append("StackTrace: \n");
                } else if (i != stackTraceElements.length - 1) {
                    buffer.append("\t |-");
                    buffer.append(stackTraceElements[i].toString());
                    buffer.append("\n");
                } else {
                    buffer.append("\t_|");
                    buffer.append(stackTraceElements[i].toString());
                }
            }
            return buffer.toString();
        }
    }
}

File: YunLog/src/main/java/com/example/yunlog/kotlin/YunLogStackTraceFormatter.kt
package com.example.yunlog.kotlin

class YunLogStackTraceFormatter : YunLogFormatter<MutableList<StackTraceElement>> {
    override fun format(data: MutableList<StackTraceElement>): String? {

        if (data.isNullOrEmpty()) {
            return null
        }

        if (data.size == 1) {
            return data[0].toString()
        }
        val logString = StringBuffer()

        for (element in data) {

            when(element){
                data.first() -> {
                    logString.append("\t StackTrace: \n").append("\t |- ").append(element.toString()).append("\n")
                }
                data.last() -> {
                    logString.append("\t |_").append(element.toString()).append("\n")
                }
                else -> {
                    logString.append("\t |- ").append(element.toString())
                }
            }
        }

        return data.toString()
    }
}